### ISSUE TITLE
High GPU Memory Usage during Batch Execution of Graphs in Graph Transformer

### ISSUE THREAD

**Issue by EloyAnguiano** [lucidrains/graph-transformer-pytorch#3]
# Description:
When running a batch of 32 graphs using the GraphTransformer object, there is a notable increase in GPU memory usage during the operation on the edges. The memory spikes from approximately 830 MiB (solely due to the CUDA context) to around 4347 MiB. The operation in question is located at line 84 in the graph_transformer_pytorch.py file in the repository.

# Problem Details:
- Dimensions of the edge tensor: torch.Size([32, 236, 236, 8]).
- Number of heads: 8 (I know that it is too much for 8-sized edges, but that number is for testing purposes only)
- Operation causing GPU memory increase: [Link to Line 84](https://github.com/lucidrains/graph-transformer-pytorch/blob/8613d2141c84f1cbaa4e2802e3091638c706c9ca/graph_transformer_pytorch/graph_transformer_pytorch.py#L84).

# Snippet
Following your example, you can launch this code to see the GPU usage:
```python
import torch
from graph_transformer_pytorch import GraphTransformer

model = GraphTransformer(
    dim = 256,
    depth = 6,
    edge_dim = 8,             # optional - if left out, edge dimensions is assumed to be the same as the node dimensions above
    with_feedforwards = True,   # whether to add a feedforward after each attention layer, suggested by literature to be needed
    gated_residual = True,      # to use the gated residual to prevent over-smoothing
    rel_pos_emb = True          # set to True if the nodes are ordered, default to False
)
device = "cuda:0"
model = model.to(device)

nodes = torch.randn(32, 236, 256, device = device)
edges = torch.randn(32, 236, 236, 8, device = device)
mask = torch.ones(32, 236, device = device).bool()

nodes, edges = model(nodes, edges, mask = mask)

```
# Optimization Proposal:
It is suggested to explore the possibility of optimizing the operation. Considering the adjacency matrix is already available, masking the edges that are irrelevant using a softmax function for masking purposes as a Transformer Decoder would do could be a potential optimization. This would eliminate the need to forward those edges, potentially reducing the GPU memory load.

While I am open to attempting these optimizations myself, my concern lies in the potential inefficiency or unintended degradation of the model if critical considerations are overlooked. Therefore, any guidance, insights, or collaboration on implementing these optimizations efficiently would be immensely valuable.

**Comment by EloyAnguiano** [comment]
Also, when I run [this line](https://github.com/lucidrains/graph-transformer-pytorch/blob/8613d2141c84f1cbaa4e2802e3091638c706c9ca/graph_transformer_pytorch/graph_transformer_pytorch.py#L86) the GPU usage goes to `7877MiB` and when I continue over [this line](https://github.com/lucidrains/graph-transformer-pytorch/blob/8613d2141c84f1cbaa4e2802e3091638c706c9ca/graph_transformer_pytorch/graph_transformer_pytorch.py#L98) it goes up to `11359MiB`.
